### PR TITLE
#21219 clean up dynamic fields

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
Adding JSON capabilities for ms-SQL 
After these changes, we should be able to support ms-SQL + progress and no values will get saved in the dynamic columns

### CODE REVIEW

**Review by nollymar — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/business/json/ContentletJsonHelper.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+     * @throws JsonProcessingException
+     */
+    public String writeAsString(final Object object) throws JsonProcessingException {
+        return objectMapper.get().writeValueAsString(object);
+    }
+
+    /**
+     * Jackson mapper configuration and lazy initialized instance.
+     */
+    private final Lazy<ObjectMapper> objectMapper = Lazy.of(() -> {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/business/json/ContentletJsonHelper.java:53]: move this to the top

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 2]
```diff
@@ -271,6 +272,47 @@ protected Object loadField(String inode, String fieldContentlet) throws DotDataE
...
 	}
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected Object loadJsonField(final String inode,
+            final com.dotcms.contenttype.model.field.Field field) throws DotDataException {
+
+        String loadJsonFieldValueSQL = null;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:282]: nice

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 3]
```diff
@@ -2082,19 +2124,28 @@ private String getInode(final String existingInode, final Contentlet contentlet)
...
+            final Map<String, Object> fieldsMap = getFieldsMap(contentlet);
+            try {
+                addDynamicFields(upsertValues, fieldsMap, "date");
+                addDynamicFields(upsertValues, fieldsMap, "text");
+                addDynamicFields(upsertValues, fieldsMap, "text_area");
+                addDynamicFields(upsertValues, fieldsMap, "integer");
+                addDynamicFields(upsertValues, fieldsMap, "float");
+                addDynamicFields(upsertValues, fieldsMap, "bool");
+            } catch (JsonProcessingException e) {
+                throw new DotDataException(e);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:2137]: Log me

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 4]
```diff
@@ -2792,20 +2864,35 @@ public Queries getQueries(final Field field, final Date maxDate) {
      */
     public Queries getJsonFieldQueries(final Field field, final Date maxDate) {
         final ContentletJsonAPI contentletJsonAPI = APILocator.getContentletJsonAPI();
-        if (!contentletJsonAPI.isPersistContentAsJson()) {
-            return new NullQueries();
+        if (contentletJsonAPI.isPersistContentAsJson()) {
+            //If we have got this far it means we are a running postgres instance that obviously supports json.
+            final SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:2869]: should be this configurable?

File: dotCMS/src/main/java/com/dotcms/integritycheckers/HostIntegrityChecker.java [hunk 5]
```diff
@@ -205,6 +222,11 @@ public boolean generateIntegrityResults(final String endpointId) throws Exceptio
                 return false;
             }
 
+             String contentAsJson = StringPool.BLANK;
+
+            if(DbConnectionFactory.isPostgres()) { contentAsJson = " or c.contentlet_as_json->'fields'->'hostName'->>'value' = ht.host"; }
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/integritycheckers/HostIntegrityChecker.java:227]: format this better

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/business/HostAPIImpl.java [hunk 6]
```diff
@@ -910,40 +912,68 @@ private void sendUnArchiveSiteSystemEvent (final Contentlet contentlet) {
...
+    private synchronized Host getOrCreateDefaultHost() throws DotDataException, DotSecurityException {
 
-            if("isDefault".equalsIgnoreCase(field.variable())){
-                isDefault = field;
-            }
+        final ContentType siteContentType = hostType();
+        final List<Field> fields = siteContentType.fields();
+        final Optional<Field> defaultField = fields.stream().filter(field -> "isDefault".equalsIgnoreCase(field.variable())).findFirst();
+
+        if(!defaultField.isPresent()){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/business/HostAPIImpl.java:921]: I would add a log here

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 7]
```diff
@@ -271,6 +272,47 @@ protected Object loadField(String inode, String fieldContentlet) throws DotDataE
 	    return m.get(fieldContentlet);
 	}
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:278]: doc

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 8]
```diff
@@ -271,6 +272,47 @@ protected Object loadField(String inode, String fieldContentlet) throws DotDataE
...
+    protected Object loadJsonField(final String inode,
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:279]: Should the return here be a Optional<String>?

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 9]
```diff
@@ -2148,6 +2199,27 @@ private void addDynamicFields(final List<Object> upsertValues, final Map<String,
         }
     }
 
+    /**
+     * Sets default values where
+     * 0 is set where there is supposed to be any numeric val,
+     * false is set where there was supposed to be a bool val and null everywhere else.
+     * @param upsertValues
+     * @param prefix
+     */
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:2208]: I think this doc need to be improved

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 10]
```diff
@@ -2792,20 +2864,35 @@ public Queries getQueries(final Field field, final Date maxDate) {
      */
     public Queries getJsonFieldQueries(final Field field, final Date maxDate) {
         final ContentletJsonAPI contentletJsonAPI = APILocator.getContentletJsonAPI();
-        if (!contentletJsonAPI.isPersistContentAsJson()) {
-            return new NullQueries();
+        if (contentletJsonAPI.isPersistContentAsJson()) {
+            //If we have got this far it means we are a running postgres instance that obviously supports json.
+            final SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+            if (DbConnectionFactory.isPostgres()) {
+                final String select = String
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:2871]: Can we have something like a Interface here and do something like

```

interface XXX {
     getUpdateQuery();
     getAnyOtherQuery();
}

PostgresXXX implements XXX{

}

MsSqlXXX implements XXX {

}

AnyOtherDataBaseXX implements XXX {
}
```

in this way we not need to keep asking by the BataBase all the time

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java [hunk 11]
```diff
@@ -304,6 +305,24 @@ public Object loadField(String inode, Field f) throws DotDataException {
         return contentFactory.loadField(inode, f.getFieldContentlet());
     }
 
+    @CloseDBIfOpened
+    public Object loadField(final String inode, final com.dotcms.contenttype.model.field.Field field) throws DotDataException{
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java:309]: Optional<Object> in the return?

File: dotCMS/src/main/java/com/dotcms/content/business/json/ContentletJsonHelper.java (hunk 1 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/business/json/ContentletJsonHelper.java:53]: done

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java (hunk 7 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:278]: but it's there it's inherited

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java (hunk 10 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:2871]: I thought about that. but there are cases on which a big SQL is already in place and we need to modify just a small fragment to reach the json field and the query gets built dynamically concatenating those fragments

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/business/HostAPIImpl.java (hunk 6 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/business/HostAPIImpl.java:921]: done

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java (hunk 8 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:279]: It does not return Optional to be compliant with the original method called loadField but I guess you're right

File: dotCMS/src/main/java/com/dotcms/integritycheckers/HostIntegrityChecker.java (hunk 5 above)
Comment (fabrizzio-dotCMS) [reply dotCMS/src/main/java/com/dotcms/integritycheckers/HostIntegrityChecker.java:227]: done

**Pull request merged**